tests/tests/test_compat.py

```diff
@@ -189,40 +189,68 @@ def assert_successful_deployment(api_deployments, deployment_id, timeout=TIMEOUT
+        self._exception_q = queue.Queue(maxsize=1)
+        self.detector = self._client.get_reboot_detector(self._host_ip)
+        self.barrier = barrier
+
+    def run(self):
+        try:
+            # Setup the individual reboot-detectors
+            with self.detector as rd:
+                # Possibly move this to after the reboot-detector accepts connections (?)
+                self.barrier.wait()
+                rd.verify_reboot_performed(max_wait=self._timeout.total_seconds())
+        except Exception as e:
+            self._exception_q.put_nowait(e)
+
+    def join(self, timeout=None):
+        super().join(timeout)
+        if not self._exception_q.empty():
+            exc = self._exception_q.get_nowait()
+            if isinstance(exc, Exception):
+                raise exc
+
+
+class ParallelRebootDetector:
     """
-    Watches for clients to reboot in separate threads
+       A parallel implementation of the RebootDetector class
+
+    :param clients: running Mender clients
+    :param host_ip: The IP of the host
+    :param timeout: The timeout for the reboot detector
+    :param deadline: The deadline for the timeout, waiting for the reboot detector(s) to finish the join
     """
 
-    class _RebootThread(threading.Thread):
-        def __init__(self, host_ip, client, timeout, **kwargs):
-            super().__init__(**kwargs)
-            self._host_ip = host_ip
-            self._client = client
-            self._timeout = timeout
-            self._exception_q = queue.Queue(maxsize=1)
-
-        def run(self):
-            try:
-                detector = self._client.get_reboot_detector(self._host_ip)
-                with detector as rd:
-                    rd.verify_reboot_performed(max_wait=self._timeout.total_seconds())
-            except Exception as e:
-                self._exception_q.put_nowait(e)
-
-        def join(self, timeout=None):
-            super().join(timeout)
-            if not self._exception_q.empty():
-                exc = self._exception_q.get_nowait()
-                if isinstance(exc, Exception):
-                    raise exc
-
-    threads = []
-    for client in clients:
-        thread = _RebootThread(host_addr, client, timeout)
-        thread.start()
-        threads.append(thread)
-    return threads
+    def __init__(self, clients, host_ip, timeout, deadline):
+        self.reboot_detectors = []
+        self.deadline = deadline
+        self.enter_barrier = threading.Barrier(
+            len(clients) + 1  # +1 for the main-thread to block
+        )
+        for client in clients:
+            self.reboot_detectors.append(
+                RebootDetectorThread(host_ip, client, timeout, self.enter_barrier)
+            )
+
+    def __enter__(self):
+        for d in self.reboot_detectors:
+            d.start()
+
+        # Wait for all reboot-detectors to finish their setup, before starting
+        # the code within the 'with' statement
+        self.enter_barrier.wait()
```

Comment:
Is it 100% certain that the parallel threads can't raise an exception? Can you add a short timeout here (~10sec) just to be safe? And also abort the test `if self.enter_barrier.broken` after.

Reply:
As far as I understand, all exceptions are caught and put into the exception queue? And can the barrier be broken if no one has called `abort`?

Reply:
I believe the barrier "breaks" if you time out waiting for it.

Reply:
> This method may raise a BrokenBarrierError exception if the barrier is broken or reset while a thread is waiting.

Is this not good enough? I mean, the test will fail with a broken barrier exception